src/main/drivers/flash_w25n01g.c

```diff
@@ -580,6 +590,209 @@ void w25n01g_pageProgramFinish(flashDevice_t *fdevice)
+
+    bufferDirty = false;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// Write operation has just completed
+busStatus_e w25n01g_callbackWriteComplete(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+
+    fdevice->currentWriteAddress += fdevice->callbackArg;
+
+    // Call transfer completion callback
+    if (fdevice->callback) {
+        fdevice->callback(fdevice->callbackArg);
+    }
+
+    return BUS_READY;
+}
+
+uint32_t w25n01g_pageProgramContinue(flashDevice_t *fdevice, uint8_t const **buffers, uint32_t *bufferSizes, uint32_t bufferCount)
+{
+    if (bufferCount < 1) {
+        fdevice->callback(0);
+        return 0;
+    }
+
+    // The segment list cannot be in automatic storage as this routine is non-blocking
+    STATIC_DMA_DATA_AUTO uint8_t readStatus[] = { W25N01G_INSTRUCTION_READ_STATUS_REG, W25N01G_STAT_REG, 0 };
+    STATIC_DMA_DATA_AUTO uint8_t readyStatus[3];
+    STATIC_DMA_DATA_AUTO uint8_t writeEnable[] = { W25N01G_INSTRUCTION_WRITE_ENABLE };
+    STATIC_DMA_DATA_AUTO uint8_t progExecCmd[] = { W25N01G_INSTRUCTION_PROGRAM_EXECUTE, 0, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progExecCmd2[] = { W25N01G_INSTRUCTION_PROGRAM_EXECUTE, 0, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progExecDataLoad[] = { W25N01G_INSTRUCTION_PROGRAM_DATA_LOAD, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progRandomProgDataLoad[] = { W25N01G_INSTRUCTION_RANDOM_PROGRAM_DATA_LOAD, 0, 0};
+
+    static busSegment_t segmentsDirty[] = {
+        {.u.buffers = {readStatus, readyStatus}, sizeof(readStatus), true, w25n01g_callbackReady},
+        {.u.buffers = {writeEnable, NULL}, sizeof(writeEnable), true, w25n01g_callbackWriteEnable},
+        {.u.buffers = {progExecCmd, NULL}, sizeof(progExecCmd), true, w25n01g_callbackProgExecute},
+        {.u.buffers = {readStatus, readyStatus}, sizeof(readStatus), true, w25n01g_callbackReady},
+        {.u.buffers = {writeEnable, NULL}, sizeof(writeEnable), true, w25n01g_callbackWriteEnable},
+        {.u.buffers = {progRandomProgDataLoad, NULL}, sizeof(progRandomProgDataLoad), false, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL},
+        {.u.link = {NULL, NULL}, 0, true, NULL},
+    };
+
+    static busSegment_t segmentsNotDirty[] = {
+        {.u.buffers = {readStatus, readyStatus}, sizeof(readStatus), true, w25n01g_callbackReady},
+        {.u.buffers = {writeEnable, NULL}, sizeof(writeEnable), true, w25n01g_callbackWriteEnable},
+        {.u.buffers = {progExecDataLoad, NULL}, sizeof(progExecDataLoad), false, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL},
+        {.u.buffers = {NULL, NULL}, 0, true, NULL},
+        {.u.link = {NULL, NULL}, 0, true, NULL},
+    };
+
+    busSegment_t *firstSegment;
+    busSegment_t *lastSegment;
+
+    // Ensure any prior DMA has completed before continuing
+    spiWait(fdevice->io.handle.dev);
+
+    if (bufferDirty) {
```

Comment:
This code does probably work, but is so hard to follow.
It may be better to build program every time. Difference in performance won't be that big, it will prevent some nasty bugs (what is some part of template was overwritten by mistake/not update correctly).

Something like:
```
busSegment_t* waitStatus(busSegment_t* s) {
    return OP_READ(s, readStatus, readyStatus, w25n01g_callbackReady);
}

busSegment_t* randomProgramDataLoad(busSegment_t* s, uint32_t programLoadAddress, void* buff, size_t len) {
    uint32_t columnAddress = W25N01G_LINEAR_TO_COLUMN(programLoadAddress);
    progRandomProgDataLoad.addr[0] = (columnAddress >> 8) & 0xff;
    progRandomProgDataLoad.addr[1] = columnAddress & 0xff;
    s = OP_WRITE(s, progRandomProgDataLoad, NULL);
    s = OP_WRITE_BUFF(s, buff, len, NULL);
    return s;
}
```

It will still be headache to get static buffer right, but it will resemble normal program flow.

Reply:
I have coded the behaviour to be exactly as before with the same logic, just making it non-blocking. It is rather complex as a result, but if that is to change I would suggest it be in a follow on PR.

Reply:
This is not to change behavior, just construction of segments sequence. Old code did use at most 2 segment sequence (in fact single transaction), so it was considerably easier to follow.
Now the logic is spread ovet multiple places (static with command sequence / segment referencing  / code to modify command / code to select/modify/extend  sequence. You can tell what does `segmentsDirty[7]` contain?

Using this generator approach, sequence would be assembled command-by-command, in normal, sequential way. Only execution would be asynchronous ...

Reply:
It’s possible to chain segments, so I’ll re-code to do that which will make the code easier to follow rather than the current patching of a single sequence.

Reply:
That's good idea, it should make code quite easy to follow and won't prevent segment refactoring in future.
(It will cost 16 bytes per link. IMO well worth code simplification)